pkg/kubelet/kubelet_server_journal.go

```diff
@@ -0,0 +1,372 @@
+// from the client. It runs with the privileges of the calling  process
+// (the kubelet) and should only be allowed to be invoked by a root user.
+type journalServer struct{}
+
+// ServeHTTP translates HTTP query parameters into arguments to be passed
+// to journalctl on the current system. It supports content-encoding of
+// gzip to reduce total content size.
+func (journalServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	var out io.Writer = w
+	args, err := newJournalArgsFromURL(req.URL.Query())
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	w.Header().Set("Content-Type", "text/plain;charset=UTF-8")
+	if req.Header.Get("Accept-Encoding") == "gzip" {
+		w.Header().Set("Content-Encoding", "gzip")
+
+		gz := gzip.NewWriter(out)
+		defer gz.Close()
+		out = gz
+	}
+	args.Copy(out)
+}
+
+// journalArgs assists in invoking the OS specific logging command.
+type journalArgs struct {
+	Since         string
+	Until         string
+	Tail          int
+	Timeout       int
+	Format        string
+	Boot          *int
+	Services      []string
+	Pattern       string
+	CaseSensitive bool
+}
+
+// newJournalArgsFromURL returns bounds checked values based on query
+// parameters. If the OS specific commands ever creates incompatible arguments,
+// this method should introduce new parameters that preserve compatibility.
+func newJournalArgsFromURL(query url.Values) (*journalArgs, error) {
+	since, err := validateDate(query.Get("since"))
+	if err != nil {
+		return nil, fmt.Errorf("parameter 'since' is invalid: %v", err)
+	}
+
+	until, err := validateDate(query.Get("until"))
+	if err != nil {
+		return nil, fmt.Errorf("parameter 'until' is invalid: %v", err)
+	}
+
+	format, err := stringInSlice(query.Get("format"), "short-precise", "json", "short", "short-unix",
+		"short-iso", "short-iso-precise", "cat", "")
+	if err != nil {
+		return nil, fmt.Errorf("parameter 'output' is invalid: %v", err)
+	}
+	if len(format) == 0 {
+		format = "short-precise"
+	}
+	services, err := safeStrings(query["service"])
+	if err != nil {
+		return nil, fmt.Errorf("parameter 'service' is invalid: %v", err)
+	}
+	var boot *int
+	if bootStr := query.Get("boot"); len(bootStr) > 0 {
+		boot, err = validIntRange(bootStr, -100, 0)
+		if err != nil {
+			return nil, fmt.Errorf("parameter 'boot' is invalid: %v", err)
+		}
+	}
+	pattern, err := safeString(query.Get("pattern"))
+	if err != nil {
+		return nil, fmt.Errorf("parameter 'pattern' is invalid: %v", err)
+	}
+
+	// All parameters loaded from the query must be thoroughly sanitized - do
+	// not pass query parameters directly to the OS specific command without
+	// limiting them as demonstrated above.
+	return &journalArgs{
+		Services:      services,
+		Since:         since,
+		Until:         until,
+		Tail:          boundedIntegerOrDefault(query.Get("tail"), 0, 100000, 0),
+		Boot:          boot,
+		Timeout:       boundedIntegerOrDefault(query.Get("timeout"), 1, 60, 30),
```

Comment:
60 second timeout limit seems a bit low... Thoughts on increasing it to 180 seconds (3 minutes)?

Reply:
sounds reasonable to me, updated.